# exerise 4.5

In [2]:
# 01. Importing Libraries
# import libraries

import pandas as pd
import numpy as np
import os

Mixed type data

In [3]:
# creating a dataframe

df_test = pd.DataFrame()

In [4]:
# create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [5]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [6]:
# check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [7]:
# fixing mixed type column

df_test['mix'] = df_test['mix'].astype('str')

Importing products.csv

In [8]:
# file path for products.csv

products_filepath = "/Users/daphne/Downloads/Instacart Basket Analysis/02 Data/Original Data/products.csv"

In [9]:
# import products.csv

df_prods = pd.read_csv(products_filepath)

 Finding missing values

In [10]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
df_prods.shape

(49693, 5)

In [14]:
#creating new dataframe without missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
df_prods_clean.shape

(49677, 5)

Duplicates

In [18]:
# finding duplicates

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
df_prods_clean.shape

(49677, 5)

In [22]:
# dropping duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [23]:
df_prods_clean_no_dups.shape

(49672, 5)

Exporting new dataframe

In [29]:
# exporting df_prods_clean_no_dups"

output_folder = "/Users/daphne/Downloads/Instacart Basket Analysis"
output_file = "df_prods_checked.csv"
output_path = f"{output_folder}/{output_file}"

In [30]:
df_prods_clean_no_dups.to_csv(output_path, index=False)

# Task 

Importing order.csv

In [31]:
# file path for orders.csv

orders_filepath = "/Users/daphne/Downloads/Instacart Basket Analysis/02 Data/Original Data/orders.csv"

In [32]:
# import orders.csv
df_ords = pd.read_csv(orders_filepath)

## Task, step 2

In [33]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Min order_dow is 0, there is no 0 day of the week. Max order_dow is 6, it should be 7. Maybe 0 repersents the first day of the week and 6 repersents the last but that doesn't make sense to me. Max order_hour_of_day is 2.3, I belive that means 23, I did get confused by the decimal though. days_since_prior_order is also confusing because why is max 3?

## Task, step 3

In [35]:
# checking for mixed type data 

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## Task, step 4

Step 4 implies that there is a column with mixed type data, but I got no output for step 3. So no, there is no column in df_ords with mixed type data.

## Task, step 5

In [36]:
# finding missing values

df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are a few senarios that come to mind, as to why there are 206,209 missing values in "days_since_prior_order" column. There are no missing values in any other column, this shows that it's porobably not a data entry error. I would assume that most of the missing values are from first time customers, as in peope who are making there first order, so there is no days since their prior order. Another senario is orders placed in the same day, if an order is placed before the next day then there are no days since their prior order. 

In [37]:
# creating new dataframe with missing values

df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [38]:
df_ords_nan.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
11   2168274        2    prior             1          2                 11   
26   1374495        3    prior             1          1                 14   
39   3343014        4    prior             1          6                 11   
45   2717275        5    prior             1          3                 12   

    days_since_prior_order  
0                      NaN  
11                     NaN  
26                     NaN  
39                     NaN  
45                     NaN

In [39]:
df_ords_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

Just out of curiosity I wanted to see the dataframe with missing values and it looks like my senario of first time cutomers was accurate. Every row with order_number = 1 has a null value under days_since_prior_order.

## Task, step 6

I will leave the missing values as is because the missing data is relavent and helpful to the dataframe. It is an indicator of a first time customer. 

## Task, step 7

In [40]:
# finding duplicates

df_ords_dups = df_ords[df_ords.duplicated()]

In [41]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

df_ords has no duplicate values.

## Task, step 8

There are no duplicate values to address.

## Task, step 9

df_prods_checked was already exported. I didn't make any changes to df_ords. 